lib/krikri/ldp/invalidatable.rb

```diff
@@ -0,0 +1,102 @@
+module Krikri::LDP
+  ##
+  # Implements invalidation for `Krikri::LDP::Resource`s. This is different 
+  # from deletion, in that the resource continues to respond `200 OK`, and 
+  # return the representation, Nothing is removed from the LDP server.
+  # 
+  # Works as a mixin to `Krikri::LDP::Resource`, assuming an implementation of
+  # `#rdf_source`, which may simply return `self`.
+  #
+  # @example invalidating a resource
+  #   class MyResource
+  #     include Krikri::LDP::Resource
+  #     include Krikri::LDP::Invalidatable
+  #
+  #     def rdf_subject
+  #       @rdf_subject ||= RDF::URI('http://example.com/ldp/a/resource/path')
+  #     end
+  #   end
+  #  
+  #   invalidatable_resource = MyResource.new
+  #   # the resource must exist before it can be invalidated! 
+  #   invalidatable_resource.save
+  #
+  #   invalidatable_resource.invalidate!
+  #   invalidatable_resource.invalidated? # => true
+  #   invalidatable_resource.invalidated_at_time 
+  #   # => Thu, 03 Dec 2015 10:27:45 -0800
+  #
+  # @see http://www.w3.org/TR/2013/REC-prov-dm-20130430/#term-Invalidation 
+  #   for documentation on PROV invalidation
+  module Invalidatable
+    # @see RDF::PROV
+    INVALIDATED_BY_URI = RDF::PROV.wasInvalidatedBy
+    INVALIDATED_TIME_URI = RDF::PROV.invalidatedAtTime
+
+    ##
+    # Invalidates the resource by marking it with a `prov:invalidatedAtTime`. If
+    # an `RDF::Term` is passed as the first argument, that term is used as the 
+    # value of `prov:wasInvalidatedBy`.
+    #
+    # @example invalidating with an activity
+    #   invalidatable_resource.invalidate!(RDF::URI('http://example.org/moomin'))
+    #   invalidatable_resource.was_invalidated_by
+    #   # => #<RDF::URI:0x2acab846109c URI:http://example.org/moomin>
+    #
+    # @param activity_uri [RDF::Term] a URI for the invalidating activity. If 
+    #   none is given, this defaults to `nil` and no `prov:wasInvalidatedBy`
+    #   statement is added.
+    #
+    # @raise [RuntimeError] when the resource does not exist or is already 
+    #   invalid
+    # @return [void]
+    def invalidate!(activity_uri = nil)
+      raise "Cannot invalidate #{rdf_subject}, does not exist." unless exists?
+
+      # force a reload unless we have cached an invalidatedAtTime
+      rdf_source.get({}, true) unless invalidated?
+      # we check invalidated again in case the reload came back invalid
+      raise "Cannot invalidate #{rdf_subject}, already invalid." if invalidated?
```

Comment:
I guess I see the point about the contract with the Activity to record _it_ as the invalidator and fail if that can't be done. I wish there could be some way to _warn_ the Activity that an aggregation was already invalidated, just because of the case where you might want to rerun an activity that invalidates records based on some query (without having to revise the query).  It seems that a special exception class is necessary, as you mentioned. Depending on the situation, the exception could be handled one way or another (e.g. logging and continuing).

Just out of curiosity, is there any possibility of having some kind of Activity in our system that's not one of the enqueued activities? Like a "manual run" activity?